Comments and cleanup for loop cloning

BruceForstall opened the pull request:
This is a no-diff change

Added various comments to document loop cloning.

Standardized and improved some logging.

Consolidated more loop cloning condition checking into
`optIsLoopClonable` that was previously in `optIdentifyLoopOptInfo`.

Replaced some `0` weights with `BB_ZERO_WEIGHT`.

Made FMT_BB use pervasive.

BruceForstall committed:
Comments and cleanup for loop cloning

This is a no-diff change

Added various comments to document loop cloning.

Standardized and improved some logging.

Consolidated more loop cloning condition checking into
`optIsLoopClonable` that was previously in `optIdentifyLoopOptInfo`.

Replaced some `0` weights with `BB_ZERO_WEIGHT`.

Made FMT_BB use pervasive.

BruceForstall commented:
@dotnet/jit-contrib

AndyAyersMS commented on src/coreclr/jit/loopcloning.cpp:
```diff
@@ -436,16 +441,20 @@ void LoopCloneContext::PrintBlockConditions(unsigned loopNum)
     JitExpandArrayStack<LC_Condition>& conds = *conditions[loopNum];
 
-    JITDUMP("Evaluating %d loop cloning conditions for loop %d\n", conds.Size(), loopNum);
+    JITDUMP("Evaluating %d loop cloning conditions for loop L%02u\n", conds.Size(), loopNum);
```
I see `FMT_LP` out there, knocking at the door.

kunalspathak replied:
+1

BruceForstall replied:
I almost did this, but deferred it. Will add it to the "to do" list.

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5279,20 +5378,31 @@ void Compiler::optCloneLoops()
 
-    JITDUMP("\nCloning loop %d: [h: %d, f: %d, t: %d, e: %d, b: %d, c: %d].\n", loopInd,
-            optLoopTable[loopInd].lpHead->bbNum, optLoopTable[loopInd].lpFirst->bbNum,
+    JITDUMP("\nCloning loop L%02u: [head: " FMT_BB ", first: " FMT_BB ", top: " FMT_BB ", entry: " FMT_BB
```
Maybe cache `optLoopTable[loopInd]` as you have done in other methods?

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5555,14 +5664,25 @@ BasicBlock* Compiler::optInsertLoopChoiceConditions(LoopCloneContext* context,
 }
 
+//------------------------------------------------------------------------
+// OptEnsureUniqueHead: Ensure that loop "loopInd" has a unique head block.
```
It would be nice to see us more standard terminology. But maybe save this for a subsequent PR.

In the literature, this block is usually called a `preheader` (see eg Muchnick section 7.4) and the `header` is the block in the loop that can be entered from outside the loop.

BruceForstall replied:
Not only the terminology, but also the semantics are weird. What Muchnick calls the header block we call the entry. Our head block is the lexical immediate (non-loop) predecessor of our "first" block. So is what this is creating a pre-header? Sounds like it, but our weird semantics make it hard to understand that.

BruceForstall replied:
There's also `fgCreateLoopPreHeader()`

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5082,64 +5084,89 @@ void Compiler::optPerformStaticOptimizations(unsigned loopNum, LoopCloneContext*
-    // First, for now, make sure the loop doesn't have any embedded exception handling -- I don't want to tackle
-    // inserting new EH regions in the exception table yet.
-    BasicBlock* stopAt       = optLoopTable[loopInd].lpBottom->bbNext;
+    LoopDsc& loop = optLoopTable[loopInd];
```
I don't suppose this can be `const LoopDsc&` ...?

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5166,40 +5193,110 @@ bool Compiler::optIsLoopClonable(unsigned loopInd)
         return false;
     }
 
+    if (!(((loop.lpTestOper() == GT_LT || loop.lpTestOper() == GT_LE) && (loop.lpIterOper() == GT_ADD)) ||
```
Seems like this is a prime candidate for `operIs(...)`

BruceForstall replied:
I went ahead and created a new static OperIs() which can be used to (slightly) simplify this.

AndyAyersMS commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5166,40 +5193,110 @@ bool Compiler::optIsLoopClonable(unsigned loopInd)
+        JITDUMP("Loop cloning: rejecting loop L%02u. It has %d returns;"
+                " if added to previously existing %d returns, it would exceed the limit of %d.\n",
+                loopInd, loopRetCount, fgReturnCount, epilogLimit);
+        return false;
```
Don't forget https://github.com/dotnet/runtime/issues/48740#issuecomment-785938375

BruceForstall replied:
fyi, I had previously added this to the list of loop work items: https://github.com/dotnet/runtime/issues/43549

AndyAyersMS reviewed (approved):
Left a few comments. Likely you can defer them for future PRs.

kunalspathak commented on src/coreclr/jit/codegenlinear.cpp:
```diff
@@ -346,6 +346,9 @@ void CodeGen::genCodeForBBlist()
         if ((block->bbPrev != nullptr) && (block->bbPrev->bbJumpKind == BBJ_COND) &&
             (block->bbWeight != block->bbPrev->bbWeight))
         {
+            JITDUMP("Adding label due to BB weight difference: BBJ_COND " FMT_BB " with weight " FMT_WT
```
While you are here, I am wondering if there is an easy way to add towards the end the mapping of `BB -> IG` somehow? While looking at JITDump, it is not straightforward to find this mapping and is very helpful overall when we look backward in JITDump.

kunalspathak commented on src/coreclr/jit/loopcloning.cpp:
```diff
@@ -392,36 +393,40 @@ JitExpandArrayStack<JitExpandArrayStack<LC_Condition>*>* LoopCloneContext::Ensur
     if (levelCond == nullptr || levelCond->Size() == 0)
     {
-        JITDUMP("No block conditions\n");
+        printf("No block conditions\n");
```
Any reason why these and below are made `printf` instead of `JITDUMP`?

BruceForstall replied:
JITDUMP checks verbose (is COMPlus_JitDump set); printf does not. This means a function like this can be called from the debugger even if verbose isn't set. Of course, the caller probably wants to check `verbose`. In general, if `verbose` is already checked, I prefer direct `printf`

kunalspathak replied:
Ah, didn't realize that this method is called under `verbose`.

kunalspathak commented on src/coreclr/jit/optimizer.cpp:
```diff
@@ -5439,8 +5547,9 @@ void Compiler::optCloneLoop(unsigned loopInd, LoopCloneContext* context)
     // Now go through the new blocks, remapping their jump targets within the loop.
-    for (BasicBlock* blk = optLoopTable[loopInd].lpFirst; blk != optLoopTable[loopInd].lpBottom->bbNext;
-         blk             = blk->bbNext)
+    for (BasicBlock* blk = optLoopTable[loopInd].lpFirst; //
```
remove the extra `//`  present at the end of this line and next line?

BruceForstall replied:
This is a hack to keep jit-format from merging the three lines into 2. I'm ok with either 1 line or 3, but 2 for a `for` is annoying, IMO.

kunalspathak reviewed (approved):
LGTM with minor comments.

BruceForstall commented on src/coreclr/jit/loopcloning.h:
```diff
@@ -4,34 +4,108 @@
+    3. Once the optimization candidates are identified, we derive cloning conditions.
+       For example: to clone a simple "for (i=0; i<n; ++i) { a[i] }" loop, we need the
+       following conditions:
+              (a != null) && (n >= 0) && (n <= a.length) && (stride > 0)
```
Yeah, I listed that as one of the limitations down below.

BruceForstall committed:
Review feedback

Added FMT_LP formatting string.

Cached often-used `optLoopTable[loopInd]` expression.

Added `const` to many loop query member functions.

Added static `GenTree::OperIs(compareOper, oper, oper, oper...)`
functions for simplifying oper check expressions where the
compareOper isn't from a GenTree node `OperGet()`. (This doesn't
need to be part of GenTree, but it doesn't hurt, either.)

Added a few more comments.

BruceForstall committed:
Fix static OperIs

Rename to StaticOperIs. It appears the compiler uses the wrong
template in some cases, but doesn't complain about duplicate
options, leading to run-time failures.
